File: src/syntax/token.rs
// Follows names here
// https://github.com/golang/go/blob/master/src/cmd/compile/internal/syntax/tokens.go

#[derive(Debug, Clone, PartialEq)]
pub enum Kind {
    Lparen, // (
    Rparen, // )
    Lbrack, // [
    Rbrack, // ]
    Bslash, // \
    Id,
    Int,
    Float,
    String,
    Illegal,
    Eof,
}

#[derive(Clone)]
pub struct Loc {
    pub row: usize,
    pub col: usize,
    pub pos: usize,
}

#[derive(Clone)]
pub struct Token {
    pub kind: Kind,
    pub loc: Loc,
    pub literal: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_token() {
        let tok = Token {
            kind: Kind::Lparen,
            loc: Loc {
                row: 1,
                col: 1,
                pos: 2,
            },
            literal: String::from("("),
        };
        assert_eq!(&tok.literal, "(");
    }
}

File: .arxived/parser/token/token.go
package token

import "fmt"

type TokenType uint

const (
	Lparen TokenType = iota // (
	Rparen                  // )
	Lbrack                  // [
	Rbrack                  // ]
	Bslash                  // \
	Id
	Int
	Float
	String
	Illegal
	Eof
)

type Loc struct {
	Row      uint
	Column   uint
	Position uint
}

type Token struct {
	Type    TokenType
	Literal string
	Loc     Loc
}

func (tok *Token) String() string {
	return fmt.Sprintf("token{ type: %v, loc: (%3d, %3d), literal: \"%v\" }",
		tok.Type, tok.Loc.Row, tok.Loc.Column, tok.Literal)
}

func (t TokenType) String() string {
	switch t {
	case Lparen:
		return "Lparen    "
	case Rparen:
		return "Rparen    "
	case Lbrack:
		return "Lbrack    "
	case Rbrack:
		return "Rbrack    "
	case Bslash:
		return "Bslash    "
	case Id:
		return "Id        "
	case Int:
		return "Int       "
	case Float:
		return "Float     "
	case String:
		return "String    "
	case Illegal:
		return "Illegal   "
	case Eof:
		return "Eof       "
	default:
		return fmt.Sprintf("(unknown token type: %d)", uint(t))
	}
}
